drivers/pwm/pwm-axi-pwmgen.c

```diff
@@ -0,0 +1,226 @@
+	struct regmap		*regmap;
+
+	/* Used to store the period when the channel is disabled */
+	unsigned int		ch_period[AXI_PWMGEN_NPWM];
+	bool			ch_enabled[AXI_PWMGEN_NPWM];
+};
+
+struct regmap_config axi_pwm_regmap_config = {
+	.reg_bits = 32,
+	.reg_stride = 4,
+	.val_bits = 32,
+	.max_register = AXI_PWMGEN_MAX_REGISTER,
+};
+
+static struct axi_pwmgen *to_axi_pwmgen(struct pwm_chip *chip)
+{
+	return container_of(chip, struct axi_pwmgen, chip);
+}
+
+static int axi_pwmgen_apply(struct pwm_chip *chip, struct pwm_device *device,
+			    const struct pwm_state *state)
+{
+	struct axi_pwmgen *pwm = to_axi_pwmgen(chip);
+	unsigned long clk_rate_hz = clk_get_rate(pwm->clk);
+	unsigned int ch = device->hwpwm;
+	struct regmap *regmap = pwm->regmap;
+	u64 period_cnt, duty_cnt;
+	int ret;
+
+	if (!clk_rate_hz)
+		return -EINVAL;
+
+	period_cnt = DIV_ROUND_UP_ULL(state->period * clk_rate_hz, NSEC_PER_SEC);
+	if (period_cnt > UINT_MAX)
+		return -EINVAL;
+
+	pwm->ch_period[ch] = period_cnt;
+	pwm->ch_enabled[ch] = state->enabled;
+	ret = regmap_write(regmap, AXI_PWMGEN_CHX_PERIOD(ch), state->enabled ? period_cnt : 0);
+	if (ret)
+		return ret;
+
+	duty_cnt = DIV_ROUND_UP_ULL(state->duty_cycle * clk_rate_hz, NSEC_PER_SEC);
+	ret = regmap_write(regmap, AXI_PWMGEN_CHX_DUTY(ch), duty_cnt);
+	if (ret)
+		return ret;
+
+	return regmap_write(regmap, AXI_PWMGEN_REG_CONFIG, AXI_PWMGEN_LOAD_CONFIG);
+}
+
+static int axi_pwmgen_get_state(struct pwm_chip *chip, struct pwm_device *device,
+				struct pwm_state *state)
+{
+	struct axi_pwmgen *pwm = to_axi_pwmgen(chip);
+	unsigned long clk_rate_hz = clk_get_rate(pwm->clk);
+	struct regmap *regmap = pwm->regmap;
+	unsigned int ch = device->hwpwm;
+	u32 cnt;
+	int ret;
+
+	if (!clk_rate_hz) {
+		dev_err(device->chip->dev, "axi pwm clock has no frequency\n");
+		return -EINVAL;
+	}
+
+	state->enabled = pwm->ch_enabled[ch];
+
+	if (state->enabled) {
+		ret = regmap_read(regmap, AXI_PWMGEN_CHX_PERIOD(ch), &cnt);
+		if (ret)
+			return ret;
+	} else {
+		cnt = pwm->ch_period[ch];
+	}
+
+	state->period = DIV_ROUND_CLOSEST_ULL((u64)cnt * NSEC_PER_SEC, clk_rate_hz);
+
+	ret = regmap_read(regmap, AXI_PWMGEN_CHX_DUTY(ch), &cnt);
+	if (ret)
+		return ret;
+
+	state->duty_cycle = DIV_ROUND_CLOSEST_ULL((u64)cnt * NSEC_PER_SEC, clk_rate_hz);
+
+	return 0;
+}
+
+static const struct pwm_ops axi_pwmgen_pwm_ops = {
+	.apply = axi_pwmgen_apply,
+	.get_state = axi_pwmgen_get_state,
+};
+
+static int axi_pwmgen_setup(struct axi_pwmgen *pwm, struct device *dev)
+{
+	struct regmap *regmap = pwm->regmap;
+	int idx;
+	int ret;
+	u32 val;
+
+	ret = regmap_write(regmap, AXI_PWMGEN_REG_SCRATCHPAD, AXI_PWMGEN_TEST_DATA);
+	if (ret)
+		return ret;
+
+	ret = regmap_read(regmap, AXI_PWMGEN_REG_SCRATCHPAD, &val);
+	if (ret)
+		return ret;
+
+	if (val != AXI_PWMGEN_TEST_DATA)
+		return ret;
+
+	ret = regmap_read(regmap, AXI_PWMGEN_REG_NPWM, &pwm->chip.npwm);
+	if (ret)
+		return ret;
+
+	if (pwm->chip.npwm > AXI_PWMGEN_NPWM)
+		return -EIO;
```

Comment:
This probably doesn't need to be an error, but rather a warning. We can still use 4 channels even if the IP core reports it supports more than 4 channels.
```suggestion
	if (pwm->chip.npwm > AXI_PWMGEN_NPWM) {
		dev_warn(dev, "driver is limited to %d channels\n but hardware reported %u\n",
			AXI_PWMGEN_NPWM, pwm->chip.npwm);
		pwm->chip.npwm = AXI_PWMGEN_NPWM
	}
```

Reply:
Done